cdc/log.cc

```diff
@@ -767,64 +854,21 @@ class transformer final : public change_processor {
 
     std::vector<mutation> _result_mutations;
 
-    // Keeps the next cdc$batch_seq_no value
-    int _batch_no = -1;
-    // The timestamp of changes being currently processed
-    api::timestamp_type _ts;
-    // The cdc$time value of changes being currently processed
-    bytes _tuuid;
+    std::unique_ptr<log_mutation_builder> _builder;
+
```

Comment:
Why unique_ptr instead of just member?

Reply:
Probably because log_mutation_builder is not movable due to const member field, no?

Reply:
I think it would be better if we didn't need to have a unique_ptr here.

Reply:
If const fields force us to use allocation here we should just drop them. It's sad but that's what we have with C++.

Reply:
I don't like the `unique_ptr`, but I don't want to have to move-construct / move-assign the builder either.

The problem is with how the flow is currently structured: we first call `begin_timestamp()` which initializes the state, then `process_change` which uses the state. In the call of `begin_timestamp` we have to assign `_builder`.

I was thinking about two alternative approaches:
- `begin_timestamp` returns a builder, and then the caller must pass it to `process_change`. It's nice because the builder serves as a witness (proof) to `process_change` that `begin_timestamp` was called, so we have a compile-time guarantee, we don't need to `assert`.
- have a `with_new_timestamp(callback)` function instead of `begin_timestamp`, where `callback` takes a builder as a parameter. The callback would call `process_change`. Or something.

Reply:
1. constructing builder in begin_timestamp is not worse than constructing unique_ptr to the builder
2. I agree it is error-prone to rely on the order of calls (begin_timestamp first)
3. The first of the options you presented seems more elegant but is still error-prone because a user may pass a wrong builder.
4. The second option is safer but it is also a bit cumbersome.
5. Both options expose the internals (the builder) which is not great.

Maybe option two but without passing builder to the callback would be better?

Reply:
What do you mean "without passing builder to the callback"?

`process_change` needs to get the builder somehow. It either gets it from state (so we have to overwrite the state variable, which requires a move/copy assignment operator, or there must be a pointer), or from a parameter (in which case we could pass a reference).

Reply:
I guess I was trying to encapsulate the builder inside this class. You're right that it would require assignment to a field.

I guess what I was trying to say is that:
1. assignment per se is not a problem
2. the problem is that we have no control over the order user calls member function and they can forget to call begin_timestamp

so I would either
1. just use assignment instead of unique_ptr since unique_ptr brings no value. That would leave code in a state it was before no better no worse. Which is ok for refactoring, I guess.
or
2. Think of a way to make sure that begin_timestamp is always called

The first approach you've listed does not solve it because you could do:
```
auto builder = begin_timestamp(...)
process_changes(builder)
process_changes(builder)
```

The second approach is a bit better but it leads to a more twisted code. I guess we could go with this second approach but I'm not super sure it's that much better so I would vote for just going with assignment for now until we put some more thoughts into this and maybe find some more elegant solution.

Reply:
I just found a solution.
https://en.cppreference.com/w/cpp/utility/optional/emplace